src/modm/driver/encoder/encoder_impl.hpp

```diff
@@ -0,0 +1,49 @@
+#ifndef MODM_ENCODER_HPP
+#error "Don't include this file directly, use 'encoder.hpp' instead!"
+#endif
+
+template<typename PH_A, typename PH_B, modm::encoder::Division_t DIVISION, typename UnderlyingType>
+inline void
+modm::Encoder<PH_A, PH_B, DIVISION, UnderlyingType>::update()
+{
+	static_assert(std::is_signed<UnderlyingType>::value and std::is_integral<UnderlyingType>::value,
+				  "UnderlyingType must be signed integral");
+	// Tare on power-on
+	UnderlyingType raw = 0;
+	UnderlyingType diff;
+
+	if (PH_A::read()) raw = 3;
+	if (PH_B::read()) raw ^= 1;  // convert gray to binary
```

Comment:
Separate reads may tear.
If you turn them into a `modm::SoftwareGpioPort<PH_A, PH_B>`, then you get atomic reads for free at least for pins on the same port. You can assert on the number of ports `modm::SoftwareGpioPort<PH_A, PH_B>::number_of_ports()` to force this atomicity.

Reply:
Very nice. That's my next and afterwards generalization for Divisor.

Reply:
This works, but is it sexy??

auto read = modm::platform::SoftwareGpioPort<PH_A, PH_B>::read();
if (read & 0b10) raw = 3;
if (read & 0b01) raw ^= 1;  // convert gray to binary

Reply:
> but is it sexy??

Yes, it's an atomic read and I wanna make love to it!1!!

Btw, you can do `using Phases = modm::platform::SoftwareGpioPort<PH_A, PH_B>;` in the class definition for added sexyness.

Reply:
Sweet. This beautyfied also connect()

Phases::setInput(::Gpio::InputType::PullUp);